docs/codeql/codeql-language-guides/advanced-dataflow-scenarios-cpp.rst

```diff
@@ -0,0 +1,421 @@
+.. _advanced-dataflow-scenarios-cpp:
+
+.. pull-quote:: Note
+
+   The data flow library described here is available from CodeQL 2.12.5 onwards. With the release of CodeQL 2.13.0 the library uses the new modular API for data flow. For information on the previous version of the library, see :ref:`Analyzing data flow in C and C++ <analyzing-data-flow-in-cpp>` and for information about the new modular API and how to migrate any existing queries to the updated data flow library, see `New dataflow API for CodeQL query writing <https://gh.io/codeql-new-dataflow-api>`__.
+
+Advanced dataflow scenarios for C/C++
+======================================
+
+Data flow for C and C++ distinguish between the value of a pointer and the value of what the pointer points to. We call this the "indirection" of the pointer. Tracking the pointer and its indirection as separate entities is important for precise dataflow, but it may mean that you do not get the right flow if you select the wrong data flow node. This article demonstrates various scenarios where it is important to consider whether data flow should be computed on the value of the pointer or its indirection.
+
+For almost all situations we only need to instantiate a dataflow configuration and specify our sources and sinks, and dataflow will handle everything for us.
+
+However, when a write to a field is not visible to CodeQL (for example because it happens in a function whose definition is not in the database) we need to track the qualifier, and tell the dataflow library that it is okay to transfer from flow from the qualifier to the field access. This is done by adding an ``isAdditionalFlowStep`` predicate to the dataflow module.
+
+When adding additional flow steps involving pointers one needs to be careful about whether the dataflow step should flow from the pointer or its indirection. Similarly, care must be taken to decide whether the additional step should target a pointer or its indirection.
+
+In the dual situation where a read of a field is not visible to CodeQL, the dataflow library's "implicit read" mechanism can be used to achieve the right flow.
+
+The setup
+---------
+
+Consider the following scenario: We have a two structs, each containing two fields, and we have two types of ways of getting user input, and two places that data can end up in:
+
+.. code-block:: cpp
+
+  struct A {
+    const int *p;
+    int x;
+  };
+
+  struct B {
+    A *a;
+    int y;
+  };
+
+  void sink(int);
+  int user_input();
+  void write_user_input_to(void *);
+  void read_data(const void *);
```

Comment:
This is a lot to take in, and we don't actually use all of it right away.  Can we consider moving the definitions of `A` and `B` to the block where they are first used?

Also `write_user_input_to` and `read_data` are, I think, introduced inline before they are used so are perhaps redundant here?

Reply:
I agree. In 3de32e813b5e5fa46fc9990a86b4dae8503f472d I've reworded it so that it only introduces `user_input` and `sink`. Does that improve the situation?

Reply:
Yep, I'm much happier with that.

I think when I read a code block I expect to be able to understand what is going on in it without reading ahead [far] - that's why the introduction of `A` and `B` "too early" were bothering me.

:+1: